dot/core/service_test.go

```diff
@@ -0,0 +1,651 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package core
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/network"
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/blocktree"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/keystore"
+	"github.com/ChainSafe/gossamer/lib/runtime"
+	mocksruntime "github.com/ChainSafe/gossamer/lib/runtime/mocks"
+	rtstorage "github.com/ChainSafe/gossamer/lib/runtime/storage"
+	"github.com/ChainSafe/gossamer/lib/transaction"
+	"github.com/ChainSafe/gossamer/lib/trie"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var errTestDummyError = errors.New("test dummy error")
+var testWasmPaths []string
+
+func TestGenerateWasm(t *testing.T) {
+	wasmFilePaths, err := runtime.GenerateRuntimeWasmFile()
+	require.NoError(t, err)
+	testWasmPaths = wasmFilePaths
+}
+
+func Test_Service_StorageRoot(t *testing.T) {
+	emptyTrie := trie.NewEmptyTrie()
+	ts, err := rtstorage.NewTrieState(emptyTrie)
+	require.NoError(t, err)
+	tests := []struct {
+		name         string
+		service      *Service
+		exp          common.Hash
+		retTrieState *rtstorage.TrieState
+		retErr       error
+		expErr       error
+		expErrMsg    string
+	}{
+		{
+			name:      "nil storage state",
+			service:   &Service{},
+			expErr:    ErrNilStorageState,
+			expErrMsg: ErrNilStorageState.Error(),
+		},
+		{
+			name:      "storage trie state error",
+			service:   &Service{},
+			retErr:    errTestDummyError,
+			expErr:    errTestDummyError,
+			expErrMsg: errTestDummyError.Error(),
+		},
+		{
+			name:    "storage trie state ok",
+			service: &Service{},
+			exp: common.Hash{0x3, 0x17, 0xa, 0x2e, 0x75, 0x97, 0xb7, 0xb7, 0xe3, 0xd8, 0x4c, 0x5, 0x39, 0x1d, 0x13, 0x9a,
+				0x62, 0xb1, 0x57, 0xe7, 0x87, 0x86, 0xd8, 0xc0, 0x82, 0xf2, 0x9d, 0xcf, 0x4c, 0x11, 0x13, 0x14},
+			retTrieState: ts,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			s := tt.service
+			if tt.name != "nil storage state" {
+				ctrl := gomock.NewController(t)
+				mockStorageState := NewMockStorageState(ctrl)
+				mockStorageState.EXPECT().TrieState(nil).Return(tt.retTrieState, tt.retErr)
+				s.storageState = mockStorageState
+			}
+
+			res, err := s.StorageRoot()
+			assert.ErrorIs(t, err, tt.expErr)
+			if tt.expErr != nil {
+				assert.EqualError(t, err, tt.expErrMsg)
+			}
+			assert.Equal(t, tt.exp, res)
+		})
+	}
+}
+
+func Test_Service_handleCodeSubstitution(t *testing.T) {
+	testRuntime, err := os.ReadFile(runtime.POLKADOT_RUNTIME_FP)
```

Comment:
@timwu20 Could you elaborate on this a bit more, I don't quite understand. Doesn't the Service interface provide a runtime through the config? The only usage of wasmer.NewInstance(code, cfg) I see is in handleCodeSubstitution() which I would think we need since its a change to the runtime. Unless you are saying there is a better way of going about that?

Reply:
create a new `_handleCodeSubstitution` that takes in an extra parameter `func (code []byte, cfg *wasmer.Config) (*wasmer.Instance, error)` and have `handleCodeSubstitution` call that method instead using `wasmer.NewInstance` by default.  In your unit tests you just provide your own `*runtime.Instance`

Reply:
ohhh okay I misunderstood I see. Makes a lot of sense, will implement